ouroboros-consensus/src/Ouroboros/Storage/ChainDB/Impl/Reader.hs

```diff
@@ -268,45 +260,64 @@ instructionHelper registry fromMaybeSTM fromPure CDB{..} varReader = do
     case inImmDBOrRes of
       -- We were able to obtain the result inside the transaction as we were
       -- in the 'ReaderInMem' state.
-      Right res               -> return res
+      Right fupdate -> case blockOrHeader of
+        -- We only got the header, so we have to read the whole block if
+        -- that's requested.
+        Block  -> traverse (traverse toDeserialisableBlock)  fupdate
+        Header -> traverse (traverse toDeserialisableHeader) fupdate
       -- We were in the 'ReaderInImmDB' state or we need to switch to it.
       Left (rollState, mbImmIt) -> case rollState of
         RollForwardFrom pt -> do
           immIt <- case mbImmIt of
             Just immIt -> return immIt
             -- We were in the 'ReaderInMem' state but have to switch to the
             -- 'ReaderInImmDB' state.
-            -- COST: the block at @pt@ is read.
             Nothing    -> do
               trace $ ReaderNoLongerInMem rollState
-              ImmDB.streamBlocksAfter cdbImmDB registry pt
+              ImmDB.streamAfter cdbImmDB registry blockOrHeader pt
           rollForwardImmDB immIt pt
         RollBackTo      pt -> do
           case mbImmIt of
             Just immIt -> ImmDB.iteratorClose cdbImmDB immIt
             Nothing    -> trace $ ReaderNoLongerInMem rollState
-          -- COST: the block at @pt@ is read.
-          immIt' <- ImmDB.streamBlocksAfter cdbImmDB registry pt
+          immIt' <- ImmDB.streamAfter cdbImmDB registry blockOrHeader pt
           let readerState' = ReaderInImmDB (RollForwardFrom pt) immIt'
           atomically $ writeTVar varReader readerState'
-          return $ fromPure $ RollBack pt
+          return $ pure $ RollBack pt
   where
     trace = traceWith (contramap TraceReaderEvent cdbTracer)
 
-    nextBlock :: ImmDB.Iterator (HeaderHash blk) m blk -> m (Maybe blk)
-    nextBlock immIt = ImmDB.iteratorNext cdbImmDB immIt <&> \case
-      ImmDB.IteratorResult _ _ blk -> Just blk
-      ImmDB.IteratorEBB    _ _ blk -> Just blk
-      ImmDB.IteratorExhausted      -> Nothing
-
-    rollForwardImmDB :: ImmDB.Iterator (HeaderHash blk) m blk -> Point blk
-                     -> m r
-    rollForwardImmDB immIt pt = nextBlock immIt >>= \case
-      Just blk -> do
-        let pt'          = blockPoint blk
+    toDeserialisableHeader :: Header blk -> m (Deserialisable m blk (Header blk))
+    toDeserialisableHeader hdr = return Deserialisable
+      { serialised         = Serialised $ toLazyByteString (encodeHeader hdr)
```

Comment:
This needs justification: we are always constructing the serialized form, even if the user might not necessarily care about it. This is cheap for us, because that serialized form just the annotation; however, that's then something that should be documented at the API level (docstring for "encode header" should say: we assume that this is O(1) -- maybe even rename to "get encoding" rather than "encode"), with a comment that if that assumption cannot be satisfied, we should rethink think this here -- after all, as you point out, the blocks/headers near the tip of the chain are the ones that we will read most.

Reply:
(If we are streaming stuff near the tip of the chain, and this assumption cannot be satisfied, perhaps we should change e.g. the representation of the tip of the chain to also includes serialized forms -- not needed now, but we should document assumptions and rationale at least.)